waspc/src/Wasp/Analyzer/TypeChecker/TypeError.hs

```diff
@@ -43,11 +57,47 @@ data TypeCoerceReason
   | -- | There is no relationship between the types in the coercion
     ReasonUncoercable
   | -- | A coercion of the type contained in a list failed
-    ReasonList TypeError
+    ReasonList e
   | -- | A coercion failed because a dictionary was missing a key
     ReasonDictNoKey String
   | -- | A coercion failed because a dictionary contained an extra key
     ReasonDictExtraKey String
   | -- | A coercion failed because two dictionaries had uncoercable types for a key
-    ReasonDictWrongKeyType String TypeError
+    ReasonDictWrongKeyType String e
   deriving (Eq, Show)
+
+getTypeCoercionErrorMessageAndCtx :: (Type -> TypedExpr -> String) -> TypeCoercionError -> (String, Ctx)
+getTypeCoercionErrorMessageAndCtx getUncoercableTypesMsg (TypeCoercionError (WithCtx ctx texpr) t reason) =
+  case reason of
+    ReasonList e ->
+      first (("For list element:\n" ++) . indent 2) $
+        getTypeCoercionErrorMessageAndCtx getUncoercableTypesMsg e
+    ReasonDictWrongKeyType key e ->
+      first ((("For dictionary field '" ++ key ++ "':\n") ++) . indent 2) $
+        getTypeCoercionErrorMessageAndCtx getUncoercableTypesMsg e
+    ReasonDictNoKey key -> ("Missing required dictionary field '" ++ key ++ "'", ctx)
+    ReasonDictExtraKey key -> ("Unexpected dictionary field '" ++ key ++ "'", ctx)
+    ReasonDecl -> uncoercableTypesMsgAndCtx
+    ReasonEnum -> uncoercableTypesMsgAndCtx
+    ReasonUncoercable -> uncoercableTypesMsgAndCtx
+  where
+    uncoercableTypesMsgAndCtx = (getUncoercableTypesMsg t texpr, ctx)
+
+getUnificationErrorMessageAndCtx :: TypeCoercionError -> (String, Ctx)
+getUnificationErrorMessageAndCtx = getTypeCoercionErrorMessageAndCtx $
+  \t texpr ->
+    intercalate
+      "\n"
+      [ "Can't mix following types:",
```

Comment:
```suggestion
      [ "Can't mix the following types:",
```
Should it be like this? I am also not sure, but pls check with Grammarly or sth. Grammarly for coding would be cool :D

Reply:
Grammarly says you are right, thanks :)!